### ISSUE TITLE
New rule: use type parameter for .reduce instead of a cast

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#1604]
**Repro**

```ts
const names = ['a', 'b'];

const record = names.reduce(
  (accum, name) => ({
    ...accum,
    [name]: true,
  }),
  {} as Record<string, boolean>
);
```

**Expected Result**

~Typo! It should be `Record<string, boolean>`, but because we're using a cast, it's missed.~ Ah, this does fire a TS error, thanks - updated the sample.

~Also, u~ Using a cast is unnecessary because `Array#reduce` can take in a type parameter indicating the default value & return type. The above code would be better written as:

```ts
const names = ['a', 'b'];

const record = names.reduce<Record<string, boolean>>(
  (accum, name) => ({
    ...accum,
    [name]: true,
  }),
  {} 
);
```

**Actual Result**

No error.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.7.0` |
| `@typescript-eslint/parser`        | `2.7.0` |
| `TypeScript`                       | `3.7.3` |
| `ESLint`                           | `5.15.1` |
| `node`                             | `12.4.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
Your example fires a TS error, so that does get caught.
![image](https://user-images.githubusercontent.com/7462525/74459497-41fa0800-4e40-11ea-8746-e01c24736561.png)

Though I agree - I much prefer using the type parameter. Sounds like a good stylistic rule.
